rules.go

```diff
@@ -0,0 +1,218 @@
+package main
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+)
+
+var (
+	binaryOptions = []string{
+		"document",
+		"domain",
+		"elemhide",
+		"font",
+		"genericblock",
+		"generichide",
+		"image",
+		"matchcase",
+		"media",
+		"object",
+		"other",
+		"ping",
+		"popup",
+		"script",
+		"stylesheet",
+		"subdocument",
+		"thirdparty",
+		"webrtc",
+		"websocket",
+		"xmlhttprequest",
+	}
+	// optionsSplitPat    = fmt.Sprintf(",(?=~?(?:%v))", strings.Join(binaryOptions, "|"))
+	optionsSplitPat    = fmt.Sprintf(",(~?(?:%v))", strings.Join(binaryOptions, "|"))
+	optionsSplitRe     = regexp.MustCompile(optionsSplitPat)
+	escapeSpecialRegxp = regexp.MustCompile(`([.$+?{}()\[\]\\])`)
+)
+
+// Structs
+
+type Request struct {
+	// parsed full URL of the request
+	URL *url.URL
+	// a value of Origin header
+	Origin string
+	// a value of Referer header
+	Referer string
+	// Defines is request looks like XHLHttpRequest
+	IsXHR bool
+}
+
+type Rule struct {
+	raw         string
+	ruleText    string
+	rawOptions  []string
+	regexString string
+	options     map[string]bool
+	optionsKeys []string
+	isComment   bool
+	isHTMLRule  bool
+	isException bool
+	document    bool
+	domain      map[string]bool
+}
+
+// Interfaces
+
+type RuleSet interface {
+	Allow(*Request) bool
+}
+
+// Allow methods
+
+func (rule *Rule) Allow(url string) bool {
+	// TODO
+	return false
```

Comment:
I'm sure we do not need this method. If we had it in initial specification, it was an error

Reply:
It wasn't, just Allow for RuleSet, will remove later